File: Data Structures/Practice_in_Python/Stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:
    def __init__(self):
        self.size = 0
        self.first = None
        self.last = None

    def push(self, value):
        newNode = Node(value)
        if self.size == 0:
            self.first = newNode
            self.last = self.first
        else:
            newNode.next = self.first
            self.first = newNode
        self.size += 1
        return self

    def pop(self):
        oldTop = self.first
        newTop = oldTop.next
        self.first = newTop
        oldTop.next = None
        self.size -= 1
        return oldTop.value

    def log(self):
        data = []
        current = self.first
        for _ in range(self.size):
            data.append(current.value)
            current = current.next
        print(data)


s = Stack()
s.push(10)
s.push(20)
s.push(30)
print(s.pop())
print(s.pop())
print(s.pop())
s.log()

File: Fundamentals/Data Structures/stack.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

/**
 STACK => LIFO Data Structure
 =========================== 
 

 
 Time Complexity
 ===============
 Insertion - O(1)

 Removal - O(1)

 Searching - O(N)

 Access - O(N)

 */
class Stack {
  constructor() {
    this.size = 0;
    this.first = null;
    this.last = null;
  }
  push(value) {
    let newNode = new Node(value);
    if (!this.first) {
      this.first = newNode;
      this.last = this.first;
    } else {
      let temp = newNode;
      temp.next = this.first;
      this.first = temp;
    }
    this.size++;
    return this.size;
  }
  pop() {
    if (this.size === 0) return null;
    let removed = this.first;
    if (this.size === 1) {
      this.first = null;
      this.last = null;
    } else {
      this.first = this.first.next;
    }
    this.size--;
    return removed.value;
  }
  print() {
    let arr = [];
    let temp = this.first;
    for (let i = 0; i < this.size; i++) {
      arr.push(temp.value);
      temp = temp.next;
    }
    console.log(arr);
  }
}

let stack = new Stack();
stack.push(1);
stack.push(2);
stack.push(3);
stack.push(4);
console.log(stack.push(5));
console.log(stack.pop());
// stack.pop()
// stack.pop()
// stack.pop()
// console.log(stack.pop())
stack.print();
// console.log(stack);